docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,428 @@
+  // the token denomination to be transferred
+  string denom = 1;
+  // the token amount to be transferred
+  uint64 amount = 2;
+  // coins denomination trace for tracking the source
+  DenomTrace denom_trace = 3;
+  // the sender address
+  string sender = 4;
+  // the recipient address on the destination chain
+  string receiver = 5;
+}
+```
+
+The `MsgTransfer` will validate that the `Coin` denomination from the `Token` field contains a valid hash:
+
+```golang
+func (msg MsgTransfer) ValidateBasic() error {
+  // ...
+  if err := msg.Trace.Validate(); err != nil {
+    return err
+  }
+  // Only validate the ibc denomination when trace info is not provided
+  if msg.Trace.Trace != "" {
+    denomTraceHash, err := ValidateIBCDenom(msg.Token.Denom)
+    if err != nil {
+      return err
+    }
+    traceHash := msg.Trace.Hash()
+    if !bytes.Equal(traceHash.Bytes(), denomTraceHash.Bytes()) {
+      return Errorf("token denomination trace hash mismatch, expected %s got %s", traceHash, denomTraceHash)
+    }
+  } else if msg.Trace.BaseDenom != msg.Amount[0].Denom {
+    // otherwise, validate that denominations are equal
+    return Wrapf(
+      ErrInvalidDenomForTransfer,
+      "token denom must match the trace base denom (%s ≠ %s)",
+      msg.Amount, msg.Trace.BaseDenom,
+    )
+  }
+  // ...
+}
+```
+
+```golang
+// ValidateIBCDenom checks that the denomination for an IBC fungible token is valid. It returns the hash of denomination on success.
+func ValidateIBCDenom(rawDenom string) (tmbytes.HexBytes, error) {
+  denomSplit := strings.SplitN(denom, "/", 2)
+
+  switch {
+    case denomSplit[0] == denom:
+      err = Wrapf(ErrInvalidDenomForTransfer, "denomination should be prefixed with the format 'ibc/{hash(trace + \"/\" + %s)}'", denom)
+    case denomSplit[0] != "ibc":
+      err = Wrapf(ErrInvalidDenomForTransfer, "denomination %s must be prefixed with 'ibc'", denom)
+    case len(denomSplit) == 2:
+      err = tmtypes.ValidateHash([]byte(denomSplit[1]))
+    default:
+      err = Wrap(ErrInvalidDenomForTransfer, denom)
+  }
+
+  if err != nil {
+    return nil, err
+  }
+
+  return denomSplit[1], nil
+}
+```
+
+When a fungible token is sent to from a source chain, the trace information needs to be stored with the
+new port and channel identifiers:
+
+```golang
+// SendTransfer
+// ...
+
+// NOTE: SendTransfer simply sends the denomination as it exists on its own
+// chain inside the packet data. The receiving chain will perform denom
+// prefixing as necessary.
+
+if types.SenderChainIsSource(sourcePort, sourceChannel, token.Denom) {
+  // create the escrow address for the tokens
+  escrowAddress := types.GetEscrowAddress(sourcePort, sourceChannel)
+
+  // escrow source tokens. It fails if balance insufficient.
+  if err := k.bankKeeper.SendCoins(
+    ctx, sender, escrowAddress, sdk.NewCoins(token),
+  ); err != nil {
+    return err
+  }
+} else {
+  // set the value to the lookup table if not stored already
```

Comment:
so this would only happen if we are sending a native token. 

The flow I'm thinking of is upon a recv we set the lookup if it doesn't exist. Basically when you add the prefix you set the lookup and metadata.

We don't need to set a lookup for a native token because on a refund you don't use the prefixed denom at all.